### PR TITLE
[RDY] Bytetrack

### PR DESCRIPTION
**Pull request by vigoux**
This is a mergeable version of bfredl's #12015.
As he notes in his PR, most cases are missing but this actually still solves some issue, notably with `~`.
This should also enhance treesitter's performances.

### CODE REVIEW

**Comment by teto** [comment]
do you run some benchmark ? if yes could you share. Would be interesting to compare the different situations (vim vs nvim) + check for regressions.

**Comment by vigoux** [comment]
I am copying some tiny benchmarks that I did, nothing too fancy and representative though, just to have a grasp of this.
Those benchmarks were made with hyperfine:

Small edits:
```
Benchmark #1: ./bench.sh small master
  Time (mean ± σ):      2.181 s ±  0.017 s    [User: 1.483 s, System: 0.217 s]
  Range (min … max):    2.158 s …  2.209 s    10 runs

Benchmark #2: ./bench.sh small bytetrack
  Time (mean ± σ):      1.964 s ±  0.013 s    [User: 1.462 s, System: 0.219 s]
  Range (min … max):    1.933 s …  1.978 s    10 runs

Summary
  './bench.sh bytetrack' ran
    1.11 ± 0.01 times faster than './bench.sh master'
```

Whole line edits:
```
Benchmark #1: ./bench.sh linewise master
  Time (mean ± σ):      3.093 s ±  0.019 s    [User: 1.963 s, System: 0.221 s]
  Range (min … max):    3.059 s …  3.121 s    10 runs

Benchmark #2: ./bench.sh linewise bytetrack
  Time (mean ± σ):      2.348 s ±  0.021 s    [User: 1.906 s, System: 0.219 s]
  Range (min … max):    2.316 s …  2.379 s    10 runs

Summary
  './bench.sh bigger bytetrack' ran
    1.32 ± 0.01 times faster than './bench.sh bigger master'
```

Multiline edits:
```
Benchmark #1: ./bench.sh bigger master
  Time (mean ± σ):     10.308 s ±  0.124 s    [User: 7.451 s, System: 0.340 s]
  Range (min … max):   10.193 s … 10.616 s    10 runs

Benchmark #2: ./bench.sh bigger bytetrack
  Time (mean ± σ):     10.287 s ±  0.091 s    [User: 7.408 s, System: 0.336 s]
  Range (min … max):   10.190 s … 10.511 s    10 runs

Summary
  './bench.sh bigger bytetrack' ran
    1.00 ± 0.01 times faster than './bench.sh bigger master'
```

What I conclude: bytetrack seems faster than current master by quite an extent, but those benchmarks are everything appart accurate. The explanation is that using byte-level updates allow treesitter to only incrementally parse that single byte that it needed to parse, but not the whole line, causing a significant improvement on small edits.

One important thing to note too is that this PR alos **fixes** some actual **bugs** on master, which is actually pretty important.

**Comment by norcalli** [comment]
@vigoux is this just an updated/rebased version or did you make changes to the code? Also which bugs/issues are those? They should probably be documented here and issues closed if any exist.

**Comment by vigoux** [comment]
No change to the code despite lint fixes, and it at least fixes this : https://github.com/neovim/neovim/issues/11944

**Comment by teto** [comment]
could you fix the failing tests ?

**Comment by vigoux** [comment]
I think I will close this PR, this is more WIP than I expected and I really need the help of @bfredl on this.
Fixing the various bugs will be done in later PR (hopefully) because I now found where the bug belongs.

**Comment by teto** [comment]
just linking https://github.com/neovim/neovim/pull/12646 not to forget some comments

**Comment by norcalli** [comment]
Yeah I probably wouldn't merge something that bfredl left as WIP. I was planning to try to pick up on some of his MANY MANY PRs eventually, but each one will take a little time to understand. Also normal line updates are kind of broken https://github.com/neovim/neovim/pull/11380 https://github.com/neovim/neovim/issues/11591 so we probably need to just add a shit ton of tests first and then make sure it's correct as we work on updates.

**Comment by vigoux** [comment]
Regarding invalid buffer updates, I did a bit of work in https://github.com/neovim/neovim/pull/12646 but some more work is needed. I really think that the `old_byte_size` mecanism (which treesitter uses) needs a good refactor too.

**Comment by bfredl** [comment]
@vigoux Thanks for taking care of my PR :) I hope to have time to review your changes soon, and get a basic version of this merged to master. Then we can fix more stuff + add tests on a case-by-case basis later.

**Comment by vigoux** [comment]
To be honest I haven't done much things in this pr, but in others

**Comment by vigoux** [comment]
So, a test fails, but I think the error is not actually one, let me explain why.

In the highlighting test, there is one test that in summary does the following

## The test today

Start from this
```c
void func(void) {
  abc();
  int zero = 0;
}
```

Go to this, that is, ad the end of a C block comment.

```c
void func(void) {
  abc();
  */
  int zero = 0;
}
```

This is to test that we correctly handle syntax errors (be careful, I'll use error for both syntax errors and highlighting errors, because I am lazy).

In today's implementation, the error will be found in on the `*/` because we reparse _the whole line each time_ (twice that is, because we change the buffer for both `*` and `/`), and treesitter does its thing and finds what __we__ think is the valid error node.

## What bytetrack changes

Now take the same editing, and keep in mind we parse only changed _bytes_ and not _lines_.

We hit `*`, treesitter tries hard, and thinks we are inserting a new pointer dereferencing statement (which is the best possible match here), same as before for now (except we don't reparse the indentation but, it doesn't matter here, the C parser ignores whitespaces i think).

Now we hit `/` and treesitter parses only this `/`, and not the whole line !
Treesitter thought we were inserting a pointer deref node, so it parses the `/` with that in mind, finds that this is not an expected token, goes on in a `ERROR` node, parses the `int`, still weird because nothing can fit there, parses the `zero` then the `=` and that's where it thinks he is good, because he has something like this now : `* (ERROR) zero =` which is what he thought it was !

## What we should do ?

Change the test if @bfredl agrees on this

**Comment by bfredl** [comment]
@vigoux (already talked on chat, but for the rest of the followers). I think making the highlights completely correct will be easier in the luahl branch and moving back to using luahl as we used to. Then we can setup profiling and measure if luahl makes things too slow.  So I will try to make CI pass and any last minute hot fixes, and then merge this.

**Comment by bfredl** [comment]
Also testing is insufficient, but I hope to soon make public an utility that will make it _very easy_ for ppl that want to test master to find and report errors in bytetracking.

File: src/nvim/change.c [hunk 1]
```diff
@@ -1611,6 +1611,7 @@ int open_line(
     }
     ml_replace(curwin->w_cursor.lnum, p_extra, true);
     changed_bytes(curwin->w_cursor.lnum, 0);
+    // TODO(vigoux): extmark_splice_cols here??
```
Comment (vigoux) [inline src/nvim/change.c:1614]: So, to splice or not to splice ?

File: src/nvim/extmark.h [hunk 2]
```diff
@@ -20,15 +20,21 @@ typedef struct
 
 typedef kvec_t(ExtmarkInfo) ExtmarkInfoArray;
 
+// BAD, NO GOOD name, something we can :s/// later
+typedef ptrdiff_t bcount_t;
```
Comment (vigoux) [inline src/nvim/extmark.h:24]: That's actually a good name.

File: src/nvim/fold.c [hunk 3]
```diff
@@ -1669,7 +1669,7 @@ static void foldAddMarker(
     }
     ml_replace_buf(buf, lnum, newline, false);
     if (added) {
-      extmark_splice_cols(buf, (int)lnum-1, (int)line_len,
+      extmark_splice_cols(curbuf, (int)lnum-1, (int)line_len,
```
Comment (vigoux) [inline src/nvim/fold.c:1672]: Looks like it is a rebase error from me here.

File: src/nvim/fold.c [hunk 4]
```diff
@@ -1742,10 +1742,9 @@ static void foldDelMarker(
       assert(p >= line);
       memcpy(newline, line, (size_t)(p - line));
       STRCPY(newline + (p - line), p + len);
-      ml_replace_buf(buf, lnum, newline, false);
-      extmark_splice_cols(buf, (int)lnum-1, (int)(p - line),
-                          (int)len,
-                          0, kExtmarkUndo);
+      ml_replace(lnum, newline, false);
+      extmark_splice_cols(curbuf, (int)lnum-1, (int)(p - line),
+                          (int)len, 0, kExtmarkUndo);
```
Comment (vigoux) [inline src/nvim/fold.c:1747]: Rebase error here too.

File: test/functional/lua/buffer_updates_spec.lua [hunk 5]
```diff
@@ -236,3 +242,68 @@ describe('lua: buffer event callbacks', function()
   end)
 
 end)
+
+describe('lua buffer event callbacks: on_lines', function()
```
Comment (vigoux) [inline test/functional/lua/buffer_updates_spec.lua:246]: I think you meant
```suggestion
describe('lua buffer event callbacks: on_bytes', function()
```

File: src/nvim/change.c (hunk 1 above)
Comment (bfredl) [reply src/nvim/change.c:1614]: lets revisit this when we write the test for this case.

File: src/nvim/extmark.h (hunk 2 above)
Comment (bfredl) [reply src/nvim/extmark.h:24]: if you say so.

File: src/nvim/fold.c [hunk 6]
```diff
@@ -1742,10 +1742,9 @@ static void foldDelMarker(
       assert(p >= line);
       memcpy(newline, line, (size_t)(p - line));
       STRCPY(newline + (p - line), p + len);
-      ml_replace_buf(buf, lnum, newline, false);
+      ml_replace(lnum, newline, false);
       extmark_splice_cols(buf, (int)lnum-1, (int)(p - line),
-                          (int)len,
-                          0, kExtmarkUndo);
+                          (int)len, 0, kExtmarkUndo);
```
Comment (vigoux) [inline src/nvim/fold.c:1747]: Rebase error here too ?

Comment (bfredl) [reply src/nvim/fold.c:1747]: hmm what is the problem

Comment (vigoux) [reply src/nvim/fold.c:1747]: `ml_replace` should be `ml_replace_buf` sorry for the bad range comment.

Comment (bfredl) [reply src/nvim/fold.c:1747]: onto it

**Review by vigoux — COMMENTED** [review]
LGTM, can't approve it though, because I commited things in here.

**Comment by bfredl** [comment]
fudged the CI once more. Hoping for the green lights this time.

**Pull request merged**